pkg/autoscaler/autoscaler.go

```diff
@@ -0,0 +1,295 @@
+/*
+Copyright 2021 Cortex Labs, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package autoscaler
+
+import (
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/cortexlabs/cortex/pkg/lib/cron"
+	"github.com/cortexlabs/cortex/pkg/lib/errors"
+	libmath "github.com/cortexlabs/cortex/pkg/lib/math"
+	"github.com/cortexlabs/cortex/pkg/lib/telemetry"
+	libtime "github.com/cortexlabs/cortex/pkg/lib/time"
+	"github.com/cortexlabs/cortex/pkg/types/spec"
+	"github.com/cortexlabs/cortex/pkg/types/userconfig"
+	"go.uber.org/zap"
+)
+
+const (
+	_prometheusQueryTimeoutSeconds = 10
+	_awakenStabilizationPeriod     = 10 * time.Minute
+)
+
+type Autoscaler struct {
+	sync.Mutex
+	logger                    *zap.SugaredLogger
+	crons                     map[string]cron.Cron
+	scalers                   map[userconfig.Kind]Scaler
+	lastAwakenTimestamp       map[string]time.Time
+	awakenStabilizationPeriod time.Duration
+}
+
+func New(logger *zap.SugaredLogger) *Autoscaler {
+	return &Autoscaler{
+		logger:                    logger,
+		crons:                     make(map[string]cron.Cron),
+		scalers:                   make(map[userconfig.Kind]Scaler),
+		lastAwakenTimestamp:       make(map[string]time.Time),
+		awakenStabilizationPeriod: _awakenStabilizationPeriod,
+	}
+}
+
+func (a *Autoscaler) AddScaler(scaler Scaler, kind userconfig.Kind) {
+	a.scalers[kind] = scaler
+}
+
+func (a *Autoscaler) Awaken(api userconfig.Resource) error {
+	a.Lock()
```

Comment:
Remove the use of awaken timestamp, instead lock the function, get the deployment, and check the current value of requested replicas